File: server/src/productScopeQuery.ts
import type { Request } from "express";
import {
  sqlWhereManualProductTitleAndSeller,
  sqlWhereProductKey,
  sqlWhereTitleMatchesCanonical,
} from "./sql/articleSameProductTitle.js";

export type ProductScopeMode = "canonical" | "title" | "key";

export type ParsedProductScope = {
  /** true si hay `productTitle` o `productKey` en la query (alcance acotado). */
  manual: boolean;
  /** true si viene `productKey` (prioridad sobre título). */
  byProductKey: boolean;
  productKey: string | null;
  productTitle: string;
  sellerOrNull: string | null;
};

/** `productTitle` / `productKey` + opcional `seller` en query: acotan analíticas / informe / listados. */
export function parseProductScopeQuery(req: Request): ParsedProductScope {
  const productKey =
    typeof req.query.productKey === "string" ? req.query.productKey.trim() : "";
  const productTitle =
    typeof req.query.productTitle === "string" ? req.query.productTitle.trim() : "";
  const sellerRaw = typeof req.query.seller === "string" ? req.query.seller.trim() : "";
  const byProductKey = productKey.length > 0;
  return {
    byProductKey,
    productKey: byProductKey ? productKey : null,
    productTitle,
    sellerOrNull: sellerRaw.length > 0 ? sellerRaw : null,
    manual: byProductKey || productTitle.length > 0,
  };
}

export function productScopeMode(pq: ParsedProductScope): ProductScopeMode {
  if (pq.byProductKey && pq.productKey) return "key";
  if (pq.productTitle.trim().length > 0) return "title";
  return "canonical";
}

type ResultsAlias = "r" | "results";

/**
 * Filtro SQL sobre filas de `results` según alcance (canónico / título manual / `product_key`).
 * Los índices son posiciones de placeholder `$` en la consulta final (mismo orden que `scopeParamsAfterArticle`).
 */
export function sqlWhereProductScope(
  pq: ParsedProductScope,
  alias: ResultsAlias,
  idx: { title: number; seller: number; key: number },
): string {
  switch (productScopeMode(pq)) {
    case "key":
      return sqlWhereProductKey(alias, idx.key);
    case "title":
      return `(${sqlWhereManualProductTitleAndSeller(alias, idx.title, idx.seller)})`;
    default:
      return `(${sqlWhereTitleMatchesCanonical(alias)})`;
  }
}

/** Parámetros que siguen a `$1` = `article_id` en rutas típicas (p. ej. título+seller o `productKey`). */
export function scopeParamsAfterArticleId(pq: ParsedProductScope): unknown[] {
  switch (productScopeMode(pq)) {
    case "key":
      return [pq.productKey ?? ""];
    case "title":
      return [pq.productTitle, pq.sellerOrNull];
    default:
      return [];
  }
}

File: server-spring/src/main/java/com/compraverificada/api/sql/ProductScopeQuery.java
package com.compraverificada.api.sql;

import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;

/**
 * Equivalente a {@code server/src/productScopeQuery.ts} del backend Node.
 * Centraliza el parsing de {@code productKey} / {@code productTitle} / {@code seller}
 * que llega por query params y la composición del WHERE SQL.
 */
public final class ProductScopeQuery {

    public enum Mode { CANONICAL, TITLE, KEY }

    public final boolean manual;
    public final boolean byProductKey;
    public final String productKey;   // null si no aplica
    public final String productTitle; // "" si no aplica
    public final String sellerOrNull; // null si no aplica

    private ProductScopeQuery(boolean manual, boolean byKey, String productKey,
                              String productTitle, String sellerOrNull) {
        this.manual = manual;
        this.byProductKey = byKey;
        this.productKey = productKey;
        this.productTitle = productTitle;
        this.sellerOrNull = sellerOrNull;
    }

    public static ProductScopeQuery parse(String productKeyRaw, String productTitleRaw, String sellerRaw) {
        String productKey = productKeyRaw == null ? "" : productKeyRaw.trim();
        String productTitle = productTitleRaw == null ? "" : productTitleRaw.trim();
        String seller = sellerRaw == null ? "" : sellerRaw.trim();
        boolean byKey = !productKey.isEmpty();
        boolean manual = byKey || !productTitle.isEmpty();
        return new ProductScopeQuery(
                manual,
                byKey,
                byKey ? productKey : null,
                productTitle,
                seller.isEmpty() ? null : seller
        );
    }

    public Mode mode() {
        if (byProductKey && productKey != null) return Mode.KEY;
        if (productTitle != null && !productTitle.trim().isEmpty()) return Mode.TITLE;
        return Mode.CANONICAL;
    }

    /**
     * Filtro SQL sobre filas de {@code alias} ({@code "r"} o {@code "results"}) según el modo:
     * canónico (CTE), título manual o {@code product_key}.
     */
    public String whereProductScope(String alias) {
        switch (mode()) {
            case KEY:
                return SqlSnippets.whereProductKey(alias, "productKey");
            case TITLE:
                return "(" + SqlSnippets.whereManualProductTitleAndSeller(
                        alias, "productTitle", "seller") + ")";
            default:
                return "(" + SqlSnippets.whereTitleMatchesCanonical(alias) + ")";
        }
    }

    /**
     * Agrega los parámetros que dependen del modo. Asume que {@code articleId} ya está
     * cargado por el caller (los demás parámetros — limit, offset, days — también).
     */
    public void addScopeParams(MapSqlParameterSource params) {
        switch (mode()) {
            case KEY:
                params.addValue("productKey", productKey == null ? "" : productKey);
                break;
            case TITLE:
                params.addValue("productTitle", productTitle);
                params.addValue("seller", sellerOrNull);
                break;
            default:
                break;
        }
    }
}
